Cmap enh

wronk opened the pull request:
Here's my WIP for updating the way a user specifies the colormap (as per #1754). I also made a clarification in the plot_source_estimate docstring. @dengemann, I left my modifications to mne_analyze_colormap so that anyone can potentially pass it all 6 control points (3 neg, 3 pos) and create an asymmetric cmap. However, I took that out of the default options in the source estimate plotting functions to make the API simpler.

## Current API
Currently, user must know the appropriate float values for the cmap control points or use the 5, 10, 15 flim defaults.
`colormap = mne.viz.mne_analyze_colormap(limits=[5, 10, 15])`
`brain = stc.plot('fsaverage', 'inflated', 'rh', colormap)`
`brain.scale_data_colormap(fmin=-15, fmid=0, fmax=15, transparent=False)`

## WIP API
User can now use percentiles (in addition to hard values as before) to set the cmap control points. The old API can still be used. User can also specify to have the cmap control points set automatically based on default percentiles.

#### Replicating old behavior: fmin/fmid/fmax input with mne_analyze cmap
`fmin, fmid, fmax = 3, 4, 5`
`colormap = mne.viz.mne_analyze_colormap(limits=[fmin, fmid, fmax])`
`p = stc.plot(colormap=colormap, fmin=-1 * fmax, fmid=0, fmax=fmax, transparent=False)`
![retrotest](https://cloud.githubusercontent.com/assets/5296721/6086767/27fc493c-adf9-11e4-89b7-169e1694bc6c.jpg)


#### New behavior: Automatic limits using mne_analyze cmap
`limits = 'auto'`
`colormap = 'mne_analyze'`
`p = stc.plot(colormap=colormap, limits=limits)`
![autolimits](https://cloud.githubusercontent.com/assets/5296721/6086996/21ecd7d0-adfb-11e4-9a57-a1ee79e8b2b8.jpg)


#### New behavior: f limits with tuple and mne_analyze cmap
`colormap='mne_analyze'`
`fmin, fmid, fmax = 3, 4, 5`
`limits = (fmin, fmid, fmax)`
`p = stc.plot(colormap=colormap, limits=limits)`

#### New behavior: using f limits  with dict (specifying hard values) and mne_analyze cmap
Note the the limits are specified with keyword `pos_lims` so the user knows these will be mirrored across zero to create the negative counterparts to each limit.
`colormap='mne_analyze'`
`fmin, fmid, fmax = 3, 4, 5`
`limits = dict(kind='value', pos_lims=[fmin, fmid, fmax])`
`p = stc.plot(colormap=colormap, limits=limits)`

#### New behavior: using f limits but with dict (specifying percentiles) and mne_analyze cmap
`colormap='mne_analyze'`
`limits = dict(kind='percent', pos_lims=[90., 95., 99.])`
`p = stc.plot(colormap=colormap, limits=limits)`

#### New behavior: using f limits but with dict (specifying percentiles) and hot cmap
Note that the `lims` keyword specifies the limits here (instead of `pos_lims`) because we aren't using the mirrored mne_analyze cmap
`colormap='hot'`
`limits = dict(kind='percent', lims=[90., 95., 99.])`
`p = stc.plot(colormap=colormap, limits=limits)`
![hotlimpercent](https://cloud.githubusercontent.com/assets/5296721/6087499/0e1eee4a-ae01-11e4-8cfe-f64807af3a49.jpg)

wronk commented:
ping @agramfort @dengemann

wronk commented:
ping @agramfort @dengemann

dengemann commented on mne/source_estimate.py:
```diff
@@ -1380,11 +1380,11 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
+        fmid : float | None (Deprecated)
             The middle value on the colormap.
-        fmax : float
+        fmax : float | None (Deprecated)
```
why is this? normally we don't put it like that.  Also there is not @deprecated decorater for this function.

dengemann replied:
sorry I've been to fast. forget about this comment

dengemann replied:
I think we normally remove the doc for deprecated parameters. @agramfort @Eric89GXL what's your intuition on this?

agramfort replied:
+1

old API disappears from doc

dengemann commented on mne/viz/_3d.py:
```diff
@@ -464,6 +478,78 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
+        if colormap == 'mne_analyze':
+            if limits == 'auto':
+                ctrl_pts = np.percentile(np.abs(stc.data), [97, 98.5, 99.9])
+                colormap = mne_analyze_colormap(ctrl_pts)
```
are you sure this is a good choice? does it produce the images you shared above? If we end up with a collection of scattered sparse islands of activation this is maybe not what we want.

wronk replied:
In retrospect, I picked a bad piece of data to use. There are muscle artifacts in that time window that I didn't realize until later. I just wanted to show it's similarity to the old API.

Is there a good set of included test set data with MNE that would be good to use?

dengemann replied:
use the SPM faces data set and the auditory evoked dataset for acquiring some evidence, see related examples

dengemann commented on mne/viz/_3d.py:
```diff
@@ -464,6 +478,78 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
+                                  ' not mne_analyze. Use limits[lims]')
+            else:
+                raise ValueError('"limits" must be "auto", tuple or dict')
+
```
this block of code looks repetitive. maybe we should have a private function that does the job and call it here with different parameters. Or am I missing something?

wronk replied:
Looks like it should be possible and would condense things. I'll give that a shot.

dengemann commented on mne/viz/utils.py:
```diff
@@ -168,15 +168,19 @@ def mne_analyze_colormap(limits=[5, 10, 15], format='mayavi'):
 
     """
     l = np.asarray(limits, dtype='float')
```
let's avoid these single letter variable names if possible. Chances are high that someone will suffer when using the python debugger.

dengemann commented on mne/viz/utils.py:
```diff
@@ -168,15 +168,19 @@ def mne_analyze_colormap(limits=[5, 10, 15], format='mayavi'):
         from matplotlib import colors
-        l = (np.concatenate((-np.flipud(l), l)) + l[-1]) / (2 * l[-1])
+        if len(l) == 3:
+            l = (np.concatenate((-np.flipud(l), l)) + l[-1]) / (2 * l[-1])
```
some of the parenhteses are not needed here

dengemann replied:
forget about it, fine.

dengemann commented:
@wronk looks fine to me. The only thing I'm concerned about are the defaults. If the defaults look like the images you shared we should put more effort in finding other default values. We could use the 99th percentile to avoid maps that are driven by isolated peak vertices and try to find a good minimum. The mid value could be the percentile (max - min) / 2 or something like that.

wronk commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+
+    return ctrl_pts
+
+
```
I figured out how to avoid repeating code, but kept it in a helper function. IMO, it makes the `plot_source_estimate` function cleaner/easier to parse at the cost of a few extra lines. Lmk though if I should do away with the private helper function though.

wronk commented:
## New API showing 1) original flims 2) automatic percentiles, 3) custom percentiles with mne_analyze
![montage_dict_resize_mne_analyze](https://cloud.githubusercontent.com/assets/5296721/6588630/70497d6a-c751-11e4-88d6-9cfe8553bad2.png)

## New API showing 1) original flims and consistency using new API with 2) tuple, and 3) dict
![montage_original_flims_resize_mne_analyze](https://cloud.githubusercontent.com/assets/5296721/6588629/7049219e-c751-11e4-9e5d-2e88bb90dbea.png)

## New API showing 1) original flims and consistency using new API with 2) tuple, and 3) dict
![old_flims_resize_hot](https://cloud.githubusercontent.com/assets/5296721/6588631/704d77ee-c751-11e4-8f82-418cbfa1b8f5.png)

dengemann commented:
@wronk looks much better. Still a bit edgy and sparse, compared to our examples. Especially the 'hot' colormap.

wronk commented:
For first comment, do you mean the `lims` or `pos_lims` key in the `limits` dict? I supposed I could make it take a 3-tuple too...

If you're talking about the edgyness on the rightmost brains, they're all showing what happens when the user defines his/her own limits, so the values are arbitrary. This is using the sample_audvis data now.

dengemann commented:
Ok @wronk I get it. Column 2 is automatic? I think it stoll is a bit sparse. How would the SPM faces example look like? The auditory data are particular as they have sparse and high amplitude sources.

wronk commented:
@dengemann, I updated the description, see if that makes more sense. The control (old flims behavior) is the leftmost brain for each of those brain montages. Yes, col 2 is automatic (except in the last figure which is just showing consistency with the old behavior).

Sure thing, I'll try to rerun with faces data tomorrow or the next day.

dengemann commented:
@wronk ok, cool. I think your code is a good idea. We need to play a bit with the defaults across several datasets. That's potentiall fidgedty but a valuable contribution. Question to keep in mind: does this look nice for 80% of the datasets? Next step would be to compare the new behavior to the old behavior across our other source examples. If you're lazy, make sure the new is the default and build our docs + share the HTML so everyone can look.

dengemann commented:
> build our docs + share the HTML so everyone can look.

```
cd ./doc
make clean
make html_dev
```

wronk commented:
Will do, I'll let you know if I run into issues.

dengemann commented:
Issues are also good to report as we're approaching a new release. Normally
building docs is a very good way to learn about them.
What we need to look at is how nice the source imaging examples look when
using the new defaults compared to before.

2015-03-11 2:45 GMT+01:00 wronk <notifications@github.com>:

> Will do, I'll let you know if I run into issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1782#issuecomment-78185425>.
>

wronk commented:
Building the docs worked fine. I just had to pip install the 'joblib' package.

Here are all the images for each of the mne-python examples. These were all plotted by overriding the colormap control points to use limits='auto'. I modified the percentiles slightly -- low and high are 96 and 99.95% now, respectively, with the middle limit set to the average of the two values.  Remember that in the original examples, every one of these plots had the flims set manually.

Each image is in the same order as the website.

Phase Slope Index
![plot_mne_inverse_psi_visual_001](https://cloud.githubusercontent.com/assets/5296721/6629189/3061dde0-c8c9-11e4-954a-1f3ea0a4d3b0.png)

Coherence
Seed was placed in left auditory cortex, so the automatic limits is determined mostly by the high coherence there. That's why the rh coherence values are mostly below the alpha cutoff.
![coherence at 10 hz](https://cloud.githubusercontent.com/assets/5296721/6629250/bd805f9e-c8c9-11e4-9418-5c9499fff87c.png)

Decoding source space
![plot_decoding_spatio_temporal_source_001](https://cloud.githubusercontent.com/assets/5296721/6629290/2598cfda-c8ca-11e4-8920-ea97a1bff97f.png)

DICS beamformer
The stc here has two samples (so stc.data is ~7k x 2). The other frequency range (centered at 18.2 Hz) showed higher source power determined the colormap. Since stc.data as a whole determines colormap control points the lower power samples is muted.
27.3 Hz (original):
![plot_dics_source_power_001](https://cloud.githubusercontent.com/assets/5296721/6629327/5f33af44-c8ca-11e4-88d9-37fc32d0e057.png)
18.2 Hz (other frequency point in the stc):
![dics_source_power_freq_18](https://cloud.githubusercontent.com/assets/5296721/6629465/912c1206-c8cb-11e4-87c5-2af57992d0d3.png)

I didn't see any images for the cross-talk or point-spread functions, so I tried running `./examples/inverse/plot_mne_crosstalk_function.py` and `./examples/inverse/plot_mne_point_spread_function.py` manually but the error "EEG average reference is mandatory for inverse modeling." 

2 sample permutation test
![plot_cluster_stats_spatio_temporal_2samp_001](https://cloud.githubusercontent.com/assets/5296721/6629685/1938a5a4-c8ce-11e4-9be6-7e3ee8816b44.png)

permutation t-test
![plot_cluster_stats_spatio_temporal_001](https://cloud.githubusercontent.com/assets/5296721/6629699/402ea730-c8ce-11e4-922d-01ee78815f4d.png)

wronk commented:
Lmk how you like those percentiles and I can either commit them or redo the analysis after tweaking.

agramfort commented:
@mainakjas would you mind have a look?

dengemann commented:
Cool, not that bad! Although some of them look better than others. Coherence does not seem to be that nice.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+    """Private helper function to convert percentiles to control points.
+
+    Note: If using 'mne_analyze', generate cmap control points for a directly
+    mirrored cmap for simplicity. I.e., no normalization to account for a
```
i.e., not I.e.,

wronk replied:
@mainakjas, it's technically correct since I'm starting a sentence with the phrase. Thanks for bringing this up though. I'll change it since it's somewhat conspicuous and doc strings are not a place to be using technical grammar.

mainakjas replied:
oops, my bad. Indeed, you are correct :)

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+
+    Parameters
+    ----------
+    limits : str | 3-tuple | dict
```
why do you call it `limits` instead of `flim`?

wronk replied:
Neither @Eric89GXL or I could figure out where the 'f' in `flim` comes from, so we figured that was enough of a reason to go with something more intuitive. We could use `lims`, but the abbreviation didn't seem worth saving 2 letters.

mainakjas replied:
I thought that `f` stood for frequency, no? Same like `xlim`, `ylim`, `clim`, `vlim` etc.

wronk replied:
I can definitely change it to `clim` since it's a limit on for the colorbar. I'm probably just being thick-headed, but I'm not sure where frequency comes into the picture... Before, `flims` usually referred to z-score or activation values as it defined colorbar control points (since that's most often the content of an stc's data array).

mainakjas replied:
I'm good with `clim`

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+
+    Returns
+    -------
+    ctrl_pts : array
```
what is the `shape` of this array? can you add that to the doc

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+        ctrl_pts = np.percentile(np.abs(stc_data), [97, 99.9])
+        ctrl_pts.insert(1, np.average(ctrl_pts))
+    elif isinstance(limits, tuple):
+        assert len(limits) == 3, '"limits" tuple must be length 3'
```
this appears very non-standard to me. I would raise a `ValueError` or something of that sort. Maybe @Eric89GXL or @dengemann can confirm?

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -382,18 +426,13 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
-    colormap : str
-        The type of colormap to use.
+    colormap : str | array
+        Name of colormap to use or a custom look up table (an n x 4
```
what is `n`?

wronk replied:
It's the number of interpolation points in the colormap. I'll add that into the docstring.

wronk replied:
@Eric89GXL, can you confirm that it's number of interpolation points for a mayavi. I've poked around in pysurfer, but can't be 100% sure.

FYI, I just copied this over from pysurfer docstring to improve the this one. I had gotten confused on why an array could be passed earlier before realizing the different types `colormap` could take on.

Eric89GXL replied:
I think `mayavi` expects a `256x4` array, you should be able to test it out with a couple and see what works.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -464,8 +516,39 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
+        warnings.warn('Using fmin, fmid, fmax is deprecated and will be'
+                      ' removed in v0.10. Use "limits" instead.',
+                      DeprecationWarning)
+        for fi, f in enumerate(ctrl_pts):
```
what are you trying to do here? why do you need the index `fi`?

wronk replied:
I just need to fill in the flim value if they're using the old method and left any of the limits as None. It's a consequence of removing the defaults from the function header.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -464,8 +516,39 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
+        else:
+            transparent = True
+
+    ctrl_pts = [fmin, fmid, fmax]
```
ok, why is `ctrl_pts` a list here but an array somewhere else?

wronk replied:
It's a list everywhere. It's fixed it to be consistent.

mainakjas replied:
perfect!

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,56 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+    elif isinstance(limits, dict):
+        # Get appropriate key for limits if it's a dict
+        limit_key = ['lims', 'pos_lims'][colormap == 'mne_analyze']
+        assert limit_key in limits.keys(), ('"lims" OR "pos_lims" must be '
```
I think here you should raise a `KeyError`

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -415,6 +454,19 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
         View to use. See surfer.Brain().
     colorbar : bool
         If True, display colorbar on scene.
+    limits : str | 3-tuple | dict
```
do we really need all these options? Couldn't we do everything using a `dict`? To me, it would be super-confusing to a new user if not documented properly.

wronk replied:
For many users (especially on the first pass through data), it seems easier just to be able to specify `limits='auto'`, and then build up from there if needed. I think being able to pass a tuple will also make it the transition easy for people sticking with the old fmin, fmid, fmax behavior.

Could the `limits` documentation use updates to make it less confusing?

mainakjas replied:
ok, but you could still do `dict(kind='auto')`, no? Same with tuple. It could just be `dict(kind='value', lim=(fmin, fmid, fmax))` or something like that? I am not very sure if there should be two ways of doing the same thing. Of course, the documentation could be improved too. Maybe, we could link to the `mne_analyze_colormap` function using `See also`?

wronk replied:
Yea, good point. I guess I could go either way here. It's basically philosophical issue: do we want to make everyone conform to 1 API or make it more of a catch-all and leave more options open to the user?

Thoughts? @agramfort, @dengemann

agramfort replied:
no time to look in details but just remember:

```
$ python -c "import this" | grep "obvious way"
There should be one-- and preferably only one --obvious way to do it.
```

wronk replied:
I think I'll collapse the tuple functionality into the dict. Then `clim` can be `'auto'` or `dict` to make it clearer but leave the option to easily have the function determine automatic limits.

mainakjas commented:
That's it from me for now. @wronk thanks for the nice contribution! :)

wronk commented:
Changed the documentation to say percentile (and not quartile).

`ctrl_pts` is for setting up where the control points on the colormap are. I need `scale_pts` to be able to scale the colormap for a pysurfer brain object (this deals with the 0-1 vs. 0-255 difference in scaling). See the very end of the `plot_source_estimates` function.

wronk commented on mne/source_estimate.py:
```diff
@@ -1374,18 +1374,14 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
+        colormap : str | array
+            Name of colormap to use or a custom look up table. If array, must
+            be (n x 3) or (n x 4) array for with RGB or RGBA values between 0
+            and 255.
```
@mainakjas, I filled this out a little more. n can be between 2 and a long int. I don't think it's adding more than this to the docstring though because it takes some explanation and will just weigh everything down. If someone is passing custom colormaps, I assume they either know what they're doing or are willing to go through some trial and error.

wronk commented:
@dengemann @mainakjas, I believe I addressed all your concerns. Lmk if you have another round of comments.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,55 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+            raise ValueError('If clim is a dict, clim[kind] must be '
+                             ' "value" or "percent"')
+    else:
+        raise ValueError('"clim" must be "auto", tuple, or dict')
```
this needs to be changed

mainakjas commented on mne/source_estimate.py:
```diff
@@ -1405,6 +1401,20 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
             View to use. See surfer.Brain().
         colorbar : bool
             If True, display colorbar on scene.
+        clim : str | dict
```
I'd just say `'auto' | dict` since that is the only string accepted. Can you look around in the repo to check how it is documented when there is only one acceptable string?

wronk replied:
Not a bad idea, but looks like that part of the docstring is strictly types. See mne/preprocessing/ssp.py

>qrs_threshold : float | str
        Between 0 and 1. qrs detection threshold. Can also be "auto" to
        automatically choose the threshold that generates a reasonable
        number of heartbeats (40-160 beats / min). Only for ECG.

mainakjas replied:
sounds good.

mainakjas commented on mne/source_estimate.py:
```diff
@@ -1405,6 +1401,20 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
+                    Note: Only if use this if 'colormap' is not 'mne_analyze'.
+                    Left, middle, and right bound for colormap.
+                pos_lims : length 3 list, array, or tuple
+                    Note: Only if use this if 'colormap' is 'mne_analyze'.
```
Only use this if :)

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,55 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+                          transparent=None, alpha=1.0, time_viewer=False,
                           config_opts={}, subjects_dir=None, figure=None,
-                          views='lat', colorbar=True):
+                          views='lat', colorbar=True, clim=None):
```
why is the default `None` and not `auto`?

mainakjas replied:
what does `clim=None` do? Why is it not in the documentation?

wronk replied:
We have to deprecate fmin, fmid, fmax, so it's going to be `None` for at least one version. If `clim` is `None`, then the function will default to the old behavior where fmin, fmid, fmax = [5, 10, 15] (see line 537)

mainakjas replied:
okay, but I don't follow why `clim` must be `None` to deprecate. Why should we default to the old behavior? We should just throw a warning if people go for the old behavior, no?

wronk replied:
Oic. Right now, if I default to `"auto"` then anyone who doesn't realize this (and has code that uses fmin, fmid, fmax) will have their flims overridden by `clim` so their originally coded limits would be useless.

If I'm getting this, you think `clim` should default to `"auto"`, but be overridden if they use the fmin, fmid, fmax old behavior and a warning should be thrown.

Eric89GXL replied:
Whatever you do, the behavior of `stc.plot()` (no arguments supplied regarding limits) should not be changed for one version, as it constitutes an API change.

wronk replied:
@Eric89GXL, that's right and why it was originally coded that way. If no arguments at all are passed, `fmin, fmid, fmax` must default to (5, 10, 15) in this version. I think the logic needs to remain as is.

mainakjas replied:
yes, but you'll need to document everything in any case. Currently `None` is not documented.

mainakjas replied:
actually, on second thoughts: since `None` refers to the deprecation, you don't need to document it I think.

wronk commented:
Thanks for pointing out typos :)

mainakjas commented on mne/viz/utils.py:
```diff
@@ -167,43 +167,57 @@ def mne_analyze_colormap(limits=[5, 10, 15], format='mayavi'):
+    # Ensure limits is an array
+    limits = np.asarray(limits, dtype='float')
+
+    if len(limits) != 3 and len(limits) != 6:
```
so when is the length of limits 6? The docstring does not reflect this update

wronk replied:
This is a feature @dengemann wanted. It allows for custom mne-analyze colormaps that have nonsymmetric control points. We figured it was too complex to put in the standard plotting functions.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -353,12 +353,55 @@ def plot_trans(info, trans_fname='auto', subject=None, subjects_dir=None,
+        ctrl_pts.insert(1, np.average(ctrl_pts))
+    elif isinstance(clim, dict):
+        # Get appropriate key for clim if it's a dict
+        limit_key = ['lims', 'pos_lims'][colormap == 'mne_analyze']
```
we should throw a warning (or an error?) if both `lims` and `pos_lims` are defined and the colormap is `mne_analyze`.

wronk replied:
If you think it's needed, I could... It's pretty clear in the documentation and the function will choose the appropriate one if both are defined.

mainakjas commented on mne/viz/_3d.py:
```diff
@@ -415,6 +453,20 @@ def plot_source_estimates(stc, subject=None, surface='inflated', hemi='lh',
+                Note: Only use this if 'colormap' is 'mne_analyze'.
+                Left, middle, and right bound for colormap. Positive values
+                will be mirrored directly across zero during colormap
+                construction to obtain negative control points.
```
resulting in a total of 6 control points

wronk replied:
They won't be combined into 6 control points if both are defined, but I can add an error if they try to specify more/less than 3.

mainakjas replied:
yes, I see that. But it is ambiguous which of the two will be used, no?

mainakjas replied:
oops, I think I was responding to some other thread. If we use `pos_limits`, isn't it mirrored so that you have 6 points / bounds (not sure about the exact terminology)?

wronk replied:
Not sure if I'm following. Ambiguous if they need to specify (5, 10, 15) instead of (-15, 0, 15)? I can change the docstring to further underscore the fact that it needs to positive control points. The old flims only had users specify 3 positive points for the colormap, so it's sticking with the old intuition.

wronk replied:
Yes, its mirrored and a zero is added so there are actually 7 "control points." I don't think it's worth discussing that in the docstring -- it gets confusing fast.

mainakjas commented on mne/viz/utils.py:
```diff
@@ -167,43 +167,57 @@ def mne_analyze_colormap(limits=[5, 10, 15], format='mayavi'):
+        raise ValueError('limits must have 3 or 6 elements')
+    if len(limits) == 3 and any(limits < 0.):
+        raise ValueError('if 3 elements, limits must all be non-negative')
+    if any(np.diff(limits) <= 0):
```
shouldn't the limits be monotonically increasing only if the number of elements is 3?

wronk replied:
An error is thrown later on if they are not monotonically increasing

mainakjas replied:
sounds good.

mainakjas commented on mne/viz/utils.py:
```diff
@@ -167,43 +167,57 @@ def mne_analyze_colormap(limits=[5, 10, 15], format='mayavi'):
+            limits = np.concatenate((-np.flipud(limits), [0], limits)) /\
+                limits[-1]
+        else:
+            limits = [limits[:3], [0], limits[3:]]
```
sorry if my question is dumb :) so, what does this list of lists do?

wronk replied:
Thanks. Needs to be on length 7 list, updating it.

mainakjas commented:
@wronk : Don't we have to update tests + whats_new page -- that is @dengemann @Eric89GXL et?

Eric89GXL commented:
Yep, tests examples and whats_new should be updated

wronk commented:
@Eric89GXL, making the docs results in more brain plots than are on the examples section of the website. Should I also track  down the code for each of those extra brain plots and update their code as well?

Eric89GXL commented:
The rule is that no examples should use the deprecated API, so I think the answer to your question is yes

wronk commented:
Tests, whats_new, and examples all updated. I've never written tests or updated the whats_new page before, so extra scrutiny would be appreciated there.

mainakjas commented on mne/viz/tests/test_3d.py:
```diff
@@ -102,3 +103,36 @@ def test_plot_trans():
+    # Test both types of incorrect limits key (lims/pos_lims)
+    clim = dict(kind='value', lims=(5, 10, 15))
+    colormap = 'mne_analyze'
+    assert_raises(KeyError, _limits_to_control_points, clim, stc_data,
```
I would do the tests on the `plot_source_estimates` method or whatever the user calls. That way, you test the entire pipeline.

mainakjas commented on mne/viz/tests/test_3d.py:
```diff
@@ -102,3 +103,36 @@ def test_plot_trans():
+    # Test for correct clim values
+    clim['pos_lims'] = (5, 10, 15, 20)
+    colormap = 'mne_analyze'
+    assert_raises(ValueError, _limits_to_control_points, clim, stc_data,
```
when I run the tests, I see some warnings. You should catch them instead of displaying them. See other tests for how it's done.

mainakjas commented:
Ha, I love this online flake8 integration @Eric89GXL . @wronk you should fix these:

examples/stats/plot_cluster_stats_spatio_temporal.py:30:1: F401 'mne_analyze_colormap' imported but unused
examples/inverse/plot_dics_source_power.py:18:1: F401 'scoreatpercentile' imported but unused
examples/inverse/plot_dics_source_power.py:73:36: F821 undefined name 'fmin'
examples/inverse/plot_dics_source_power.py:73:47: F821 undefined name 'fmid'
examples/inverse/plot_dics_source_power.py:73:58: F821 undefined name 'fmax'
examples/connectivity/plot_mne_inverse_psi_visual.py:34:1: F401 'mne_analyze_colormap' imported but unused

mainakjas commented on mne/viz/tests/test_3d.py:
```diff
@@ -102,3 +103,36 @@ def test_plot_trans():
+
+
+def test_limits_to_control_points():
+    """Test helper function for determining control points from value or
```
this should be one line, otherwise it won't be printed correctly when you run the tests with the `verbose` option

wronk commented:
Thanks, added your suggested fixes and implemented another test -- I didn't have a check for 3 unique colorbar limits, which will happen if data lacks variability (e.g., if it's all zeros).

Btw, how can you see online pyflakes errors (sorry I didn't notice those earlier)?

wronk commented on mne/viz/tests/test_3d.py:
```diff
@@ -50,17 +51,18 @@ def test_plot_sparse_source_estimates():
-    stc_data[(np.random.rand(20) * n_verts * n_time).astype(int)] = 1
+    stc_size = stc_data.size
+    stc_data[(np.random.rand(stc_size / 20) * stc_size).astype(int)] = \
+        np.random.rand(stc_data.size / 20)
```
Need to add some random values to the data (1/20th or 5% here). Before, this just assigned 20 random indices to be 1, but this causes issues for `clim='auto'` because it can't produce 3 different color limits.

mainakjas commented:
On Sun, Mar 22, 2015 at 11:47 PM, wronk <notifications@github.com> wrote:

> Thanks, added your suggested fixes and implemented another test -- I
> didn't have a check for 3 unique colorbar limits, which will happen if data
> lacks variability (e.g., if it's all zeros).
>
> Btw, how can you see online pyflakes errors (sorry I didn't notice those
> earlier)?
>
Just go the Travis test log and you'll see what is causing the tests to
fail :)

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1782#issuecomment-84726315>.
>

Eric89GXL commented:
"make flake" is your friend, as is an editor with built in style and flake
checking

wronk commented:
Thanks, I think they're all taken care of now :)

mainakjas commented on doc/source/whats_new.rst:
```diff
@@ -151,6 +153,8 @@ API
 
    - Add `transform`, `unit` parameters to `read_montage` by `Teon Brooks`_
 
+   - Deprecated `fmin, fmid, fmax` in stc.plot and added clim by `Mark Wronkiewicz`_
```
nitpick. It should be `clim`

mainakjas commented:
did you try building the docs? Other than that, looks in pretty good shape to me!

wronk commented:
Yep, the docs produce all the examples I modified.

Almost forgot to update `.mailmap` Lmk if you guys see anything else.

wronk commented:
@dengemann, I think this is ready :)

wronk commented:
ping @dengemann @agramfort 
Can you guys take a look so this is finalized for the GSoC example?

dengemann commented on mne/source_estimate.py:
```diff
@@ -1370,18 +1370,14 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
             PySurfer version 0.4 or above.
-        colormap : str
-            The type of colormap to use.
+        colormap : str | array
```
@wronk sorry to torture you, whenever you specify an array type in a doc please adopt the folowing scheme, based on an example:

np.ndarray of float, shape(n_channels, n_times)

wronk replied:
There are a number of options... Something like this okay though?
`colormap : str | np.ndarray of int or float, shape(n_colors, 3 | 4)`

dengemann replied:
yes fine

dengemann replied:
please go over your docs and do something like this! We've never been that close!

wronk replied:
Going back and forth, but I think that's a relatively clean way to do the rest of the docs. Feel free to continue the torture though -- might as well get it right :p

agramfort replied:
everybody happy here?

wronk replied:
I think so. @dengemann?

dengemann commented on mne/source_estimate.py:
```diff
@@ -1401,6 +1397,20 @@ def plot(self, subject=None, surface='inflated', hemi='lh',
+                    Note: Only use this if 'colormap' is 'mne_analyze'.
+                    Left, middle, and right bound for colormap. Positive values
+                    will be mirrored directly across zero during colormap
+                    construction to obtain negative control points.
```
I'm not sure how this will look like when building the docs ... I think we need blank lines after the second level of indentation. @wronk you want to build the docs once more to see how the reference pages would look like?

wronk replied:
I'm not sure how to directly share html files, but here's a screenshot of the result:
![screenshot - 03312015 - 03 20 03 pm](https://cloud.githubusercontent.com/assets/5296721/6930628/c58aa29a-d7b9-11e4-9739-c88faeb0e909.png)

dengemann replied:
looks good, not?

wronk replied:
Yep, lgtm

dengemann commented:
@wronk thanks a lot for this nice contribution, something I always wanted to have ;-) For the array thing, let's open an issue and try to fix it for the entire code base.

wronk commented:
Awesome. Thank you (and @mainakjas) for the extensive help :)
